[Feature] Update utility connect script to version 3 (#17)
ArcadiaPower/utility-connect-react: README.md

> 

Reviewer: At the end of list item one I would add "('create' mode)" and at the end of list item two I would add "('update' mode)" so that when we make references to those modes further down, it's not confusing.

> The user flow is selected with the `config.scope` parameter described in the component API Reference below.

Reviewer: This is no longer true.

I would replace with:

The user flow is selected when the `UtilityConnectToken` that's eventually passed into Utility Connect is initialized by the accompanying server: if a `utility_credential_id` is provided, Utility Connect will automatically be opened in 'update' mode (where the user updates their credentials), otherwise it's opened in 'create' mode.

> Note that instantiating the Utility Connect widget requires a one-time-use OAuth access token in order to authenticate API requests. See [Creating OAuth Tokens](https://developers.arcadia.com/#operation/createApiAccessToken) for instructions. Note that the type of token to use (`grant_type`) differs if you are creating a user or updating a user -- see `config.accessToken` in the component API Reference below for details.

Reviewer: I think we're trying to avoid mentions of `OAuth`. Should we change this from `one-time-use Oauth access token` to `one-time-use authentication token`?

Reviewer: Same comment about the link in the subsequent sentence, `Creating Oauth Tokens`. Perhaps we should say `Creating Authentication Tokens`?

Reviewer: In the third sentence, I believe we took out `grant_type`. @dlopuch can you provide clarity here?

Reviewer 2: This is more so related to what I just put up: https://github.com/ArcadiaPower/arcadia-power/pull/12315 - but yeah I think we can just remove OAuth and refer to this as the `utility_connect_token`

Reviewer 3: The API documentation for the endpoint here will be added in [EN-330](https://arcadiapower.atlassian.net/browse/EN-330) (currently in process), so may make sense to wait until that is up to finalize... but for now:

```suggestion
Note that instantiating the Utility Connect widget requires a Utility Connect token in order to authenticate API requests. See [Creating Utility Connect Tokens](TKTK) for instructions. Note that an Access Token will not work for this purpose.
```

Author: Updated it to clif's version

## Config Options

| Name                  | Type     | Description                                                                                            | Options                              | Required | Default |
| --------------------- | -------- | ------------------------------------------------------------------------------------------------------ | ------------------------------------ | -------- | ------- |

> | `utilityConnectToken` | `string` | [Utility connect access token](https://developers.arcadia.com/#section/Authentication/Utility-Connect) |                                      | Yes      | none    |

Reviewer: I think we should standardize references to the Utility Connect Token to `UtilityConnectToken`.

Author: This is the param - it should be camel case here to match how you're passing it in.

Reviewer: Sorry, my comment is about the reference in the Description column. My understanding is that the column is used to provide more context, not to demonstrate the exact parameter name

## Config Options

| Name                | Type     | Description                                                                   | Options                              | Required | Default |
| ------------------- | -------- | ----------------------------------------------------------------------------- | ------------------------------------ | -------- | ------- |

> | `accessToken`       | `string` | API [OAuth token](https://developers.arcadia.com/#operation/createOAuthToken) |                                      | Yes      | none    |

Reviewer: The PRD has this config option named `utilityConnectToken`. Is there a good reason to use `accessToken` instead?

Author: Not really - I just kept it as-is on the utility-connect side.

Reviewer 2: let's call utilityConnectToken -- I think that disambiguates against the general access token

> | `env`               | `string` | API environment                                                               | `['local', 'sandbox', 'production']` | Yes      | none    |

Reviewer: Let's remove `local`. It shouldn't be in public documentation since it's only useful for us.

Author: Removed here

> | `newCredentialData` | `object` | Data used to prefill new credentials                                          | Content depends on value of `scope`  | No       | none    |

Reviewer: We should remove this string from the Options column: `Content depends on value of `scope`. We no longer have the concept of `scope`. 

Perhaps in the Description column we should note that this data is only used when creating new credentials and will otherwise throw an error if used with a UtilityConnectToken associated with exiting credentials?

| `callbacks`         | `object` | Callback functions                                                            |                                      | No       | none    |
| `uiTheme`           | `string` | UI color theme                                                                | `['light', 'dark']`                  | No       | 'light' |

### `config.newCredentialData`

> Data that would be used to pre-fill fields prior to credential submission of a new user. Only available for an access token in `create` mode.

Reviewer: We might want to remove language around `create` and `update` mode since there's no longer a 'scope` config where you pass `create` and `update`.

Author: But there is a mode that you get in the utility connect token

Author: I tweaked it a little. Let me know what you think.

Example `config.newCredentialData` when in the "creating user" flow:

```javascript
{
  zip_code: '11787';
}
```

Data is expected in the following format for the `create` scope:

| Data field | Description     | Required |
| ---------- | --------------- | -------- |
| `zip_code` | user's zip code | no       |

### `config.accessToken`

An OAuth access token to authenticate API requests. See [Creating OAuth Tokens](https://developers.arcadia.com/#operation/createOAuthToken).

Note that the type of `accessToken` to be used depends on the user flow:

- For `mode: 'create'`, the `accessToken` must have been created with `scope: 'utility_connect', grant_type: 'client_credentials'`

> - For `mode: 'update'`, the `accessToken` must have been created with `scope: 'utility_connect', grant_type: 'password', user_id: xxx`

Reviewer: Same comments as above, I think we should rename to `config.utilityConnectToken`. Also should remove the language around creating with scope or grant_type.

> - `sandbox`: This references our sandbox API. Use this in your development, staging or test environments.

Reviewer: We should remove the 'local' option here.

- `production`: This references our production API. **Only use this in your production app.**

### `config.callbacks`

Callback functions triggered at key points in the Utility Connect flow. Expects an object with key/value pairs where keys are as documented below.

> **`onCredentialsSubmitted`**: callback function that is triggered when the user has clicked the button to submit their credentials. Passes along a utility_credential_id which can optionally be used by the containing application to poll the status of credential verification via the client's backend (which has awareness of the associated client_user_id).

Reviewer: Should surround utility_credential_id with back ticks (`utility_credential_id`)

Reviewer: Should we provide more explicit guidance on the format of the params for the callback functions? For example, we previously displayed:
```
onCredentialsSubmitted: ({ utility_credential_id }) => { ... }),
```
for this endpoint rather than just the callback name.

I don't know what's more common/helpful for documentation

Author: `onCredentialsSubmitted: ({ utility_credential_id }) => { ... }),` is below this
